File: computeWassersteinDist.py
import pandas as pd 
import numpy as np
from scipy.linalg import norm
def computeWassersteinDist(src_file, tgt_file, src_weight_file, tgt_weight_file, plan_file, p = 1):

    src = pd.read_csv(src_file)
    tgt = pd.read_csv(tgt_file)
    src_weight = pd.read_csv(src_weight_file)
    tgt_weight = pd.read_csv(tgt_weight_file)
    plan = pd.read_csv(plan_file)

    # src *= src_weight.values
    # tgt *= tgt_weight.values
    assert(tgt_weight.values.sum() == src_weight.values.sum())
    assert(tgt_weight.values.sum() == plan["mass"].sum())
    total_weight = src_weight.values.sum()
    print(total_weight)

    # Clean the columns
    src["index"] = range(1, src.shape[0] + 1)
    tgt["index"] = range(1, tgt.shape[0] + 1)

    # Merge plan with src and tgt
    result = plan.merge(src, left_on = "from", right_on = "index")
    result = result.merge(tgt, left_on = "to", right_on = "index")

    # Compute the distance
    l_values = result[[col for col in result.columns if col[-1] == 'x' and "index" not in col]].values
    r_values = result[[col for col in result.columns if col[-1] == 'y' and "index" not in col]].values

    res = np.power(np.sum(np.sum(np.power(np.abs((r_values - l_values)), p), axis = 1) * result["mass"].values / total_weight), 1 / p)
   
    return res

if __name__ == "__main__":

    src_file = "src.csv"
    tgt_file = "tgt.csv"
    src_weight_file = "src_weights.csv"
    tgt_weight_file = "tgt_weights.csv"
    plan_file = "plan.csv"

    res = computeWassersteinDist(src_file, tgt_file, src_weight_file, tgt_weight_file, plan_file)    
    print(res)

File: computeWassersteinDist.r
# Compute the wassersetein distance between two distribution

args <- commandArgs(trailingOnly = TRUE)

src_file <- args[1]
tgt_file <- args[2]
src_weights_file <- args[3]
tgt_weights_file <- args[4]
experiment_dir <- args[5]
src_file <- paste(experiment_dir, src_file, sep = "/")
tgt_file <- paste(experiment_dir, tgt_file, sep = "/")
src_weights_file <- paste(experiment_dir, src_weights_file, sep = "/")
tgt_weights_file <- paste(experiment_dir, tgt_weights_file, sep = "/")
prefix <- strsplit(src_file, "/")[[1]]
prefix <- strsplit(prefix[length(prefix)], "_")[[1]][1]
warning(prefix)
library(transport)
library(MASS)

# Read in opt and non-opt
src <- read.csv(src_file)
tgt <- read.csv(tgt_file)

src_weights <- read.csv(src_weights_file)
tgt_weights <- read.csv(tgt_weights_file)
src_weights <- as.matrix(src_weights) 
src_weights <- as.numeric(src_weights)
tgt_weights <- as.matrix(tgt_weights) 
tgt_weights <- as.numeric(tgt_weights)

# Type conversion
wpp_src <- wpp(src, src_weights)
wpp_tgt <- wpp(tgt, tgt_weights)

plan <- transport(wpp_src, wpp_tgt, p = 1, method = 'networkflow', threads = 1)
res <- wasserstein(wpp_src, wpp_tgt, p = 1, tplan = plan, prob = TRUE)
print(res)

plan_file <- paste(prefix, "plan.csv", sep = "_")
plan_file <- paste(experiment_dir, plan_file, sep = "/")
warning(plan_file)
write.csv(plan, plan_file, row.names = FALSE)
